core/src/main/java/org/apache/accumulo/core/data/KeyBuilder.java

```diff
@@ -0,0 +1,278 @@
+    protected boolean deleted = false;
+
+    final public ColumnFamilyStep<T> row(final T row) {
+      this.row = row;
+      return this;
+    }
+
+    @Override
+    final public ColumnQualifierStep<T> columnFamily(final T columnFamily) {
+      this.columnFamily = columnFamily;
+      return this;
+    }
+
+    @Override
+    final public ColumnVisibilityStep<T> columnQualifier(final T columnQualifier) {
+      this.columnQualifier = columnQualifier;
+      return this;
+    }
+
+    @Override
+    final public Build<T> columnVisibility(final T columnVisibility) {
+      this.columnVisibility = columnVisibility;
+      return this;
+    }
+
+    @Override
+    final public Build<T> timestamp(long timestamp) {
+      this.timestamp = timestamp;
+      return this;
+    }
+
+    @Override
+    public Build<T> deleted(boolean deleted) {
+      this.deleted = deleted;
+      return this;
+    }
+
+    @Override
+    public Key build() {
+      return this.build(true);
+    }
+
+    @Override
+    public Build<T> column(final T columnFamily, final T columnQualifier, final T columnVisibility) {
+      return this.columnFamily(columnFamily).columnQualifier(columnQualifier).columnVisibility(columnVisibility);
+    }
+
+    @Override
+    public ColumnVisibilityStep<T> column(final T columnFamily, final T columnQualifier) {
+      return this.columnFamily(columnFamily).columnQualifier(columnQualifier);
+    }
+  }
+
+  private static class TextKeyBuilder extends AbstractKeyBuilder<Text> {
+
+    private final Text EMPTY_TEXT = new Text();
+
+    @Override
+    public Key build(boolean copyBytes) {
+      Text columnFamily = (this.columnFamily == null) ? EMPTY_TEXT : this.columnFamily;
+      Text columnQualifier = (this.columnQualifier == null) ? EMPTY_TEXT : this.columnQualifier;
+      Text columnVisibility = (this.columnVisibility == null) ? EMPTY_TEXT : this.columnVisibility;
+      return new Key(row.getBytes(), 0, row.getLength(), columnFamily.getBytes(), 0, columnFamily.getLength(),
+          columnQualifier.getBytes(), 0, columnQualifier.getLength(), columnVisibility.getBytes(), 0,
+          columnVisibility.getLength(), timestamp, deleted, copyBytes);
+    }
+  }
+
+  private static class ByteArrayKeyBuilder extends AbstractKeyBuilder<byte[]> {
+
+    @Override
+    public Key build(boolean copyBytes) {
+      byte[] columnFamily = (this.columnFamily == null) ? EMPTY_BYTES : this.columnFamily;
+      byte[] columnQualifier = (this.columnQualifier == null) ? EMPTY_BYTES : this.columnQualifier;
+      byte[] columnVisibility = (this.columnVisibility == null) ? EMPTY_BYTES : this.columnVisibility;
+      return new Key(row, columnFamily, columnQualifier, columnVisibility, timestamp, deleted, copyBytes);
+    }
+  }
+
+  private static class CharSequenceKeyBuilder extends AbstractKeyBuilder<CharSequence> {
+
+    private final Text EMPTY_TEXT = new Text();
+
+    @Override
+    public Key build(boolean copyBytes) {
```

Comment:
Does `copyBytes` change the behaviour here? Looks like you're already copying when you build the Text objects the first time.